typegen/evaluation/file_type_hints_collector.py

```diff
@@ -0,0 +1,159 @@
+            self.imports[class_name] = module_name
+        return True
+
+    def visit_ClassDef(self, cdef: cst.ClassDef) -> bool | None:
+        # Track ClassDefs to disambiguate functions from methods
+        self._scope_stack.append(cdef)
+        return True
+
+    def leave_ClassDef(self, _: cst.ClassDef) -> None:
+        self._scope_stack.pop()
+
+    def visit_FunctionDef(self, fdef: cst.FunctionDef) -> bool | None:
+        # Track assignments from Functions
+        # NOTE: this handles nested functions too, because the parent reference gets overwritten
+        # NOTE: before we start generating hints for its children
+        self._scope_stack.append(fdef)
+        return True
+
+    def leave_FunctionDef(self, fdef: cst.FunctionDef) -> None:
+        self._scope_stack.pop()
+
+    def visit_Param(self, node: cst.Param) -> bool | None:
+        if not hasattr(node, "annotation"):
+            return True
+        variable_name, line_number = self._get_name_and_line_number(node)
+        type_hint = self._get_annotation_value(node.annotation)
+        self._add_row(line_number, TraceDataCategory.FUNCTION_PARAMETER, variable_name, type_hint)
+        return True
+
+    def visit_FunctionDef_returns(self, node: cst.FunctionDef) -> None:
+        variable_name, line_number = self._get_name_and_line_number(node)
+        if node.returns:
+            type_hint = self._get_annotation_value(node.returns)
+        else:
+            type_hint = None
+        self._add_row(line_number, TraceDataCategory.FUNCTION_RETURN, variable_name, type_hint)
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool | None:
+        if not hasattr(node, "annotation"):
+            return True
+        _, line_number = self._get_name_and_line_number(node)
+
+        type_hint = self._get_annotation_value(node.annotation)
+        if isinstance(node.target, cst.Attribute):
+            category = TraceDataCategory.CLASS_MEMBER
+            variable_name = node.target.attr.value
+        elif isinstance(node.target, cst.Name):
+            category = TraceDataCategory.LOCAL_VARIABLE
+            variable_name = node.target.value
+        else:
+            raise TypeError("Unhandled case for: " + type(node.annotation.annotation).__name__)
+        self._add_row(line_number, category, variable_name, type_hint)
+        return True
+
+    def _get_name_and_line_number(self, node):
+        pos = self.get_metadata(PositionProvider, node).start
+        variable_name = None if not hasattr(node, "name") else node.name.value
+        variable_line_number = pos.line
+        return variable_name, variable_line_number
+
+    def _add_row(self, line_number: int, category: TraceDataCategory, variable_name: str | None, type_hint: str | None):
+        class_node = self._innermost_class()
+        class_name = None
+        if class_node:
+            class_name = class_node.name.value
+        function_node = self._innermost_class()
```

Comment:
Looks like a C&P error. This should probably be `self._innermost_function`. Update tests to reflect this bug

Reply:
Done